server/src/core/user/tests/user-service.test.ts

```diff
@@ -0,0 +1,124 @@
+        expectDatesAreClose(now, updatedBeneficiary.updatedAt);
+      });
+
+      test('should create beneficiary if does not already exist', async () => {
+        const now = new Date();
+        const res = await createDefaultService().nominateBeneficiary({ phone: '254711222333', nominator: 'donorMiddleman1' });
+        const createdBeneficiary = await usersColl().findOne({ _id: res._id });
+        createdBeneficiary.donors.sort((a, b) => a.localeCompare(b));
+        expect(createdBeneficiary.phone).toBe('254711222333');
+        expect(createdBeneficiary.donors).toEqual(['donor1', 'donorMiddleman1']);
+        expect(createdBeneficiary.roles).toEqual(['beneficiary']);
+        expect(createdBeneficiary.addedBy).toBe('donorMiddleman1');
+        expectDatesAreClose(now, createdBeneficiary.createdAt);
+        expectDatesAreClose(now, createdBeneficiary.updatedAt);
+      });
+    });
+  });
+
+  describe('nominateMiddleman', () => {
+    test('should add donor to the middlemanFor list of the middleman and add middleman role', async () => {
+      const now = new Date();
+      const res = await createDefaultService().nominateMiddleman({ phone: '254700222222', nominator: 'donor1' });
+      expect(res).toEqual({ _id: 'donor2', phone: '254700222222' });
+      const updatedMiddleman = await usersColl().findOne({ _id: res._id });
+      updatedMiddleman.roles.sort((a, b) => a.localeCompare(b));
+      expect(updatedMiddleman.roles).toEqual(['donor', 'middleman']);
+      expect(updatedMiddleman.middlemanFor).toEqual(['donor1']);
+      expectDatesAreClose(now, updatedMiddleman.updatedAt);
+    });
+
+    test('should create new user if does not exist', async () => {
+      const now = new Date();
+      const res = await createDefaultService().nominateMiddleman({ phone: '254766111222', nominator: 'donor1' });
+      const createdMiddleman = await usersColl().findOne({ _id: res._id });
+      expect(createdMiddleman.roles).toEqual(['middleman']);
+      expect(createdMiddleman.middlemanFor).toEqual(['donor1']);
+      expectDatesAreClose(now, createdMiddleman.createdAt);
+      expectDatesAreClose(now, createdMiddleman.updatedAt);
+      expect(createdMiddleman.phone).toBe('254766111222');
+      expect(createdMiddleman.password).toBe('');
+    });
+
+    test('should allow nominating user who is already a middleman for other donors', async () => {
+      const now = new Date();
+      const res = await createDefaultService().nominateMiddleman({ phone: '254700555555', nominator: 'donor2' });
+      expect(res._id).toBe('donorMiddleman1');
+      const updatedMiddleman = await usersColl().findOne({ _id: res._id });
+      expect(updatedMiddleman.roles).toEqual(['donor', 'middleman']);
+      expect(updatedMiddleman.middlemanFor).toEqual(['donor1', 'donor2']);
+      expectDatesAreClose(now, updatedMiddleman.updatedAt);
+    });
+
+    test('allows nominating beneficiary as middleman', async () => {
```

Comment:
Since that was still a grey area, I decided to add explicit tests for both so that our decision be intentional. So if we change it and decided that a middleman cannot be a beneficiary, then we'll have to update the functionality and the test.

Reply:
Oh, I see. OK.

Reply:
But we decided that `beneficiary` and `donor` are mutually exclusive.

Reply:
So this makes it possible for the donor to add on of his/her beneficiaries as a middleman. But it also allows the donor to add another donor as middleman. The interesting scenario is when the donor adds a user whose not already on the platform. That user would neither be a beneficiary nor a donor. We can make it so that the donor has the option to specify whether that donor should also be added as beneficiary. Or maybe we should pick a default. Or maybe we should just keep the user as middleman, but if they choose to make a donation, then a donor role gets automatically added and they can no longer be added as beneficiary. But this scenario is definitely one we should properly define before getting to the major release.